gemelli/preprocessing.py

```diff
@@ -8,117 +8,251 @@
+
+
+def rclr_matrix(M):
+    """
+    Robust clr transform helper function.
+    This function is built for mode 2 tensors,
+    also known as matrices.
+
+    Raises
+    ------
+    ValueError
+       Raises an error if any values are negative.
+    ValueError
+       Raises an error if the matrix has more than 2 dimension.
+
+    References
+    ----------
+    .. [1] V. Pawlowsky-Glahn, J. J. Egozcue,
+           R. Tolosana-Delgado (2015),
+           Modeling and Analysis of
+           Compositional Data, Wiley,
+           Chichester, UK
+
+    .. [2] C. Martino et al., A Novel Sparse
+           Compositional Technique Reveals
+           Microbial Perturbations. mSystems.
+           4 (2019), doi:10.1128/mSystems.00016-19.
+
+    Examples
+    --------
+
+    To use directly with a
+    prebuilt tensor.
+
+    >>> from numpy import np
+    >>> T_counts = np.array([[10, 14, 43],
+                             [ 41, 43, 14]])
+    >>> rclr_matrix(T_counts)
+
+    """
+
+    mat = np.atleast_2d(M)
+    if np.any(M < 0):
+        raise ValueError("Cannot have negative proportions")
+    if M.ndim > 2:
+        raise ValueError("Input matrix can only have two dimensions or less")
+
+    # closure following procedure in
+    # skbio.stats.composition.closure
+    M_log = M / M.sum(axis=1, keepdims=True)
+    # log transform before geo-mean
+    M_log = np.log(M_log.squeeze())
+    mask = [True] * np.product(M_log.shape)
+    mask = np.array(mask).reshape(M_log.shape)
+    mask[np.isfinite(M_log)] = False
+    # sum of rows (features)
+    M_rclr = np.ma.array(M_log, mask=mask)
+    # approx. geometric mean of the features
+    gm = M_rclr.mean(axis=-1, keepdims=True)
+    # subtracted to center log
+    M_rclr = (M_rclr - gm).squeeze().data
+    # ensure any missing are zero again
+    M_rclr[~np.isfinite(M_log)] = 0.0
+    return M_rclr
+
+
+class build(_BaseConstruct):
+
+    """
+    This class can both build N-mode
+    tensors from 2D dataframes given
+    a count table and mapping data.
+
+    A list of conditional measurements are
+    given that identify context measured
+    multiple times over the same subjects.
+    Additionally a set of subject IDs
+    must be provided. Any subjects that are
+    missing in a given condition are left
+    as completely zero.
 
     Parameters
     ----------
-    pseudocount : float, optional
-        A pseudocount used in the
-        case that a sample, feature,
-            and/or condition is completely
-            missing from a vector slice in
-            the tensor formed.
     table : DataFrame
         table of non-negative count data
-        rows = samples
-        columns = features
+        rows = features
+        columns = samples
     mapping : DataFrame
         mapping metadata for table
         rows = samples
-        columns = metadata categories
-    ID_col : str, int, or float
+        columns = categories
+    subjects : str, int, or float
```

Comment:
would it be good to keep with `biom-format` and call this samples?

Reply:
I don't like the term samples in this case because you can have several samples for a single subject-conditional pairing.

Reply:
got it, makes sense, carry on